File: project_euler/Problem046/Problem046.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem046() {
    primes := make(map[int]bool)
    primes[2] = true
    result := 3
    for true {
        if isPrime(result) {
            primes[result] = true
        } else {
            border := int(math.Sqrt(float64((result - 1) >> 1)))
            isGoldbach := false
            for i := 1; i <= border; i++ {
                if primes[result - (i * i << 1)] {
                    isGoldbach = true
                    break
                }
            }
            if !isGoldbach {
                fmt.Println(result)
                return
            }
        }
        result += 2
    }
}

func isPrime(value int) bool {
    if 1 != value % 6 && 5 != value % 6 {
        return 2 == value || 3 == value
    }

    border := int(math.Sqrt(float64(value)))
    for i := 5; i <= border; i += 6 {
        if 0 == value % i || 0 == value % (i + 2) {
            return false
        }
    }
    return 1 != value
}

File: project_euler/Problem046/Problem046.js
function Problem046() {
    let primes = new Set([2])
    let result = 3
    while (true) {
        if (isPrime(result)) {
            primes.add(result)
        } else {
            let border = Math.floor(Math.sqrt((result - 1) >>> 1))
            let isGoldbach = false
            for (let i = 1; i <= border; ++i) {
                if (primes.has(result - (i * i << 1))) {
                    isGoldbach = true
                    break
                }
            }
            if (!isGoldbach) {
                console.log(result)
                return
            }
        }
        result += 2
    }
}

function isPrime(value) {
    if (1 !== value % 6 && 5 !== value % 6) {
        return 2 === value || 3 === value
    }

    let border = Math.floor(Math.sqrt(value))
    for (let i = 5; i <= border; i += 6) {
        if (!(value % i) || !(value % (i + 2))) {
            return false
        }
    }
    return 1 !== value
}

module.exports = {
    Problem046 : Problem046
};
